Review all changes incl. MVP

ckunki opened the pull request.

pj-spoelders committed:
first working version of VS: SELECT FROM NUMBER(36,0)

pj-spoelders committed:
Add testsetup
Added comment
Schema must be uppercase for snowflake

pj-spoelders committed:
Changes to gitignore

pj-spoelders committed:
- cleaned up integration test class somewhat
- fixed "no database context issue" on new snowflake test account by attempting to create new database every time making the tests more resilient

pj-spoelders committed:
Fix expected types for join tests

pj-spoelders committed:
CI build: add snowflake service credentials step

pj-spoelders committed:
make things a bit easier to debug

pj-spoelders committed:
Merge https://github.com/exasol/snowflake-virtual-schema

pj-spoelders committed:
testing CI after  branching issues

pj-spoelders committed:
adding required files for PK

pj-spoelders committed:
fix version doc vert spacing

pj-spoelders committed:
Added summary

pj-spoelders committed:
Fix release date

pj-spoelders committed:
fix dependency #s (why does PK not fix  this automatically?)

pj-spoelders committed:
Added v1 of user guide

pj-spoelders committed:
minor cleanup

pj-spoelders committed:
fix link

pj-spoelders committed:
Adjustments to user guide

pj-spoelders committed:
Remove commented out code

pj-spoelders committed:
Fix table formatting and small improvement.

pj-spoelders committed:
Added different mapping and conversions for:
- timestamp with timezone -> timezone
- warning for NUMBER with precision > 36
- 'character' is an alias for varchar in snowflake so it gets converted to varchar and there's no padding

pj-spoelders committed:
fix error code config, add snowflake package

pj-spoelders committed:
add randomised db name addendum to allow for simultaneous integration test runs.

pj-spoelders committed:
Remove link to capabilities for now as  SIT not added to project yet.

pj-spoelders committed:
Add credentials step to release workflow

pj-spoelders committed:
Add release date

pj-spoelders committed:
cleanup todos + code in comments + formatting

pj-spoelders committed:
Changed release date

pj-spoelders committed:
set release date

pj-spoelders committed:
Update changes doc to make it a little less cryptic.

pj-spoelders committed:
Merge branch 'main' into mvp / License copyright update

ckunki commented on .gitattributes:
```diff
@@ -0,0 +1,13 @@
+.github/workflows/dependencies_update.yml                        linguist-generated=true
+.github/workflows/release_droid_print_quick_checksum.yml         linguist-generated=true
+.github/workflows/release_droid_upload_github_release_assets.yml linguist-generated=true
+.github/workflows/release_droid_prepare_original_checksum.yml    linguist-generated=true
```
```suggestion
.github/workflows/release.yml         linguist-generated=true
```

ckunki commented on .gitignore:
```diff
@@ -0,0 +1,44 @@
+*.flattened-pom.xml
+/bin/
+
+username.txt
+password.txt
+accountname.txt
```
See below, proposing to replace the 3 files by a single one and also being easier to consume for a Java application or junit tests: 

```suggestion
test.properties
```

pj-spoelders replied:
Done!

ckunki commented on .project-keeper.yml:
```diff
@@ -0,0 +1,50 @@
+  runnerOs: ubuntu-20.04
+  freeDiskSpace: false
+  exasolDbVersions:
+    - "7.1.25"
+    - "8.24.0"  
```
```suggestion    
    - "8.31.0"
    - "7.1.29"
```

pj-spoelders replied:
Done.

ckunki commented on .project-keeper.yml:
```diff
@@ -0,0 +1,50 @@
+            name: Configure Snowflake credentials
+            id: configure-snowflake-credentials
+            run: |
+              echo "$USERNAME" > username.txt
+              echo "$ACCOUNTNAME" > accountname.txt
+              echo "$PASSWORD" > password.txt
```
Please see here for creating file test.properties:
https://github.com/exasol/databricks-virtual-schema/blob/main/.project-keeper.yml#L58

pj-spoelders replied:
This has been altered, as requested

ckunki commented on .project-keeper.yml:
```diff
@@ -0,0 +1,50 @@
+              ACCOUNTNAME: '${{ secrets.ACCOUNTNAME }}',
+              PASSWORD: '${{ secrets.PASSWORD }}'
+            }
+    - name: release.yml
+      stepCustomizations:
+        # Configure Snowflake credentials
+        - action: INSERT_AFTER
+          stepId: verify-release
+          content:
+            name: Configure Snowflake credentials
+            id: configure-snowflake-credentials
+            run: |
+              echo "$USERNAME" > username.txt
+              echo "$ACCOUNTNAME" > accountname.txt
+              echo "$PASSWORD" > password.txt
+            env: {
+              USERNAME: '${{ secrets.USERNAME }}',
+              ACCOUNTNAME: '${{ secrets.ACCOUNTNAME }}',
+              PASSWORD: '${{ secrets.PASSWORD }}'
+            }
```
IMHO: Release build doesn't require running tests.

```suggestion
```

pj-spoelders replied:
It is the standard workflow in project-keeper AFAIK.

ckunki commented on .vscode/settings.json:
```diff
@@ -0,0 +1,23 @@
+    "vmArgs": [
+      "-Djava.util.logging.config.file=src/test/resources/logging.properties",
+      "-Dtest.debug='true'",
+      "-Dcom.exasol.dockerdb.image=8.24.0"
```
```suggestion
      "-Dcom.exasol.dockerdb.image=8.31.0"
```

ckunki commented on README.md:
```diff
@@ -0,0 +1,46 @@
+[![Lines of Code](https://sonarcloud.io/api/project_badges/measure?project=com.exasol%3Asnowflake-virtual-schema&metric=ncloc)](https://sonarcloud.io/dashboard?id=com.exasol%3Asnowflake-virtual-schema)
+
+# Overview
+
```
```suggestion
Snowflake Virtual Schema (short "VSSF") is an implementation of a [Virtual Schema](https://docs.exasol.com/db/latest/database_concepts/virtual_schemas.htm).

```

pj-spoelders replied:
I'll add it

ckunki commented on README.md:
```diff
@@ -0,0 +1,46 @@
+
+The **Snowflake Virtual Schema** provides an abstraction layer that makes an external [Snowflake](https://www.Snowflake.org/) database accessible from an Exasol database through regular SQL commands. The contents of the external Snowflake database are mapped to virtual tables which look like and can be queried as any regular Exasol table.
+
+If you want to set up a Virtual Schema for a different database system, please head over to the [Virtual Schemas Repository][virtual-schemas].
```
Link is broken.

ckunki commented on README.md:
```diff
@@ -0,0 +1,46 @@
+* [Changelog](doc/changes/changelog.md)
+* [Dependencies](dependencies.md)
+
+Find all the documentation in the [Virtual Schemas project][vs-doc].
```
Link is broken.

pj-spoelders replied:
Fixing it

ckunki commented on doc/design.md:
If file is empty, then please delete it.

pj-spoelders replied:
ok

ckunki commented on doc/developers_guide/developers_guide.md:
Please use name singular "Developer Guide" and file name `developer_guide.md`.

Please add documentation for integration tests, e.g. does the developer need a snow flake account? etc.

pj-spoelders replied:
Done.

ckunki commented on doc/user_guide/snowflake_user_guide.md:
```diff
@@ -0,0 +1,143 @@
+
+[Snowflake](https://www.snowflake.com/) operates a platform that provides data storage via cloud computing and allows for data analysis.
+
+## Uploading the JDBC Driver to Exasol BucketFS
```
Could you please add a sentence explaining, why uploading something to somewhere is required?
Maybe a section like "Installation" would help?
Proposal add a preview about the required steps, e.g. 1. upload driver, 2. .... and add "see the following sections for details".

pj-spoelders replied:
This documentation is quasi identical to that of the other VS, so I'd rather not reinvent the wheel here.

ckunki commented on doc/user_guide/snowflake_user_guide.md:
```diff
@@ -0,0 +1,143 @@
+
+1. Download the [SnowflakeJDBC driver](https://docs.snowflake.com/en/developer-guide/jdbc/jdbc-download).
+
+2. Upload the driver to BucketFS, see [BucketFS documentation](https://docs.exasol.com/db/latest/administration/on-premise/bucketfs/accessfiles.htm).
+
+   Hint: Put the driver into folder `default/drivers/jdbc/` to register it for [ExaLoader](#registering-the-jdbc-driver-for-exaloader), too.
```
```suggestion
2. Upload the driver to BucketFS under `default/drivers/jdbc/snowflake/`, see [BucketFS documentation](https://docs.exasol.com/db/latest/administration/on-premise/bucketfs/accessfiles.htm).
```

pj-spoelders replied:
Again, as stated above. Let's not reinvent the wheel here, at this point in time.

ckunki commented on doc/user_guide/snowflake_user_guide.md:
```diff
@@ -0,0 +1,143 @@
+## Registering the JDBC driver for ExaLoader
+
+In order to enable the ExaLoader to fetch data from the external database you must register the driver for ExaLoader as described in the [Installation procedure for JDBC drivers](https://github.com/exasol/docker-db/#installing-custom-jdbc-drivers).
+1. ExaLoader expects the driver in BucketFS folder `default/drivers/jdbc`.<br />
+   If you uploaded the driver for UDF to a different folder, then you need to [upload](#uploading-the-jdbc-driver-to-exasol-bucketfs) the driver again.
+2. Additionally, you need to create file `settings.cfg` and [upload](#uploading-the-jdbc-driver-to-exasol-bucketfs) it to the same folder in BucketFS. Contents below:
```
```suggestion
Create file `settings.cfg` and [upload](#uploading-the-jdbc-driver-to-exasol-bucketfs) it to folder `default/drivers/jdbc/snowflake/` in BucketFS, too. Contents below:
```

pj-spoelders replied:
Not going to touch this.

ckunki commented on doc/user_guide/snowflake_user_guide.md:
````diff
@@ -0,0 +1,143 @@
+NOSECURITY=YES
+
+```
+Make sure there's a newline at the end of the `settings.cfg` file, as shown above, or it will not be properly read out, the EXALoader will display an error message.
````
```suggestion
Make sure there's an empty line at the end of the `settings.cfg` file, as shown above, or it will not be properly read out, the EXALoader will display an error message.
```

pj-spoelders replied:
Ok, will change this

ckunki commented on doc/user_guide/snowflake_user_guide.md:
```diff
@@ -0,0 +1,143 @@
+CREATE OR REPLACE JAVA ADAPTER SCRIPT ADAPTER.SNOWFLAKE_JDBC_ADAPTER AS
+  %scriptclass com.exasol.adapter.RequestDispatcher;
+  %jar /buckets/<BFS service>/<bucket>/virtual-schema-dist-12.0.0-snowflake-0.1.0.jar;
+  %jar /buckets/<BFS service>/<bucket>/drivers/jdbc/snowflake-jdbc-<snowflake-driver-version>.jar;
```
```suggestion
  %jar /buckets/<BFS service>/<bucket>/drivers/jdbc/snowflake/snowflake-jdbc-<snowflake-driver-version>.jar;
```

pj-spoelders replied:
That's not how it works if you follow the steps above. So won't change this.

ckunki commented on doc/user_guide/snowflake_user_guide.md:
```diff
@@ -0,0 +1,143 @@
+
+## Defining a Named Connection
+
+Define the connection to the Snowflake database as shown below. We recommend using TLS to secure the connection.
+
```
I couldn't find TLS in the connection. If using TLS is the default, then I propose to remove the sentence.

```suggestion
Define the connection to the Snowflake database as shown below.
```

pj-spoelders replied:
Thought I did this but apparently I didn't. I'll remove this.

ckunki commented on doc/user_guide/snowflake_user_guide.md:
````diff
@@ -0,0 +1,143 @@
+IDENTIFIED BY '<password>';
+```
+
+| Variable | Description |
+|----------|-------------|
+
+| `<account name` | Account name or 'account identifier' of the Snowflake platform account. |
````
```suggestion
| Variable | Description |
|----------|-------------|
| `<account name>` | Account name or 'account identifier' of the Snowflake platform account. |
```

pj-spoelders replied:
Done

ckunki commented on doc/user_guide/snowflake_user_guide.md:
```diff
@@ -0,0 +1,143 @@
+|----------|-------------|
+
+| `<account name` | Account name or 'account identifier' of the Snowflake platform account. |
+
```
Proposal: Add additional rows to the table explaining  '<user>' and '<password>', and from where to retrieve those or how to configure snowflake in order to use these credentials for a virtual schema.

pj-spoelders replied:
The detailed explanation is right below the table ...

ckunki commented on doc/user_guide/snowflake_user_guide.md:
```diff
@@ -0,0 +1,143 @@
+
+## Creating a Virtual Schema
+
+Use the following SQL command in Exasol database to create a Snowflake Virtual Schema 
```
```suggestion
Use the following SQL statement in Exasol database to create a Snowflake Virtual Schema 
```

ckunki commented on doc/user_guide/snowflake_user_guide.md:
```diff
@@ -0,0 +1,143 @@
+| `<schema name>`         | Name of the database schema you want to use in the Snowflake database. Please use UPPERCASE.        |
+
+
+For additional parameters coming from the base library see also [Adapter Properties for JDBC-Based Virtual Schemas](https://github.com/exasol/virtual-schema-common-jdbc#adapter-properties-for-jdbc-based-virtual-schemas).
```
```suggestion
For additional parameters see also [Adapter Properties for JDBC-Based Virtual Schemas](https://github.com/exasol/virtual-schema-common-jdbc#adapter-properties-for-jdbc-based-virtual-schemas).
```

pj-spoelders replied:
That's better, I agree.

ckunki commented on doc/user_guide/snowflake_user_guide.md:
```diff
@@ -0,0 +1,143 @@
+| POLYGON                  | ✓            | VARCHAR(2000000)           |                                                                                                                                                  |
+| REAL                     | ✓            | DOUBLE                     |                                                                                                                                                  |
+| SMALLINT                 | ✓            | DECIMAL(5,0)               |                                                                                                                                                  |
+| SMALLSERIAL              | ? (untested) |                            |                                                                                                                                                  |
+| SERIAL                   | ? (untested) |                            |                                                                                                                                                  |
```
```suggestion
| SMALLSERIAL              | - | | (unsupported) | 
| SERIAL                          | - | | (unsupported) |
```

pj-spoelders replied:
I'll revise this whole table.

pj-spoelders replied:
Revised.

ckunki commented on pom.xml:
```diff
@@ -0,0 +1,190 @@
+            <artifactId>error-reporting-java</artifactId>
+            <version>1.0.1</version>
+        </dependency>
+        <dependency>
+            <groupId>net.snowflake</groupId>
+            <artifactId>snowflake-jdbc</artifactId>
+            <version>3.16.1</version>
+        </dependency>
```
```suggestion
        <dependency>
            <groupId>net.snowflake</groupId>
            <artifactId>snowflake-jdbc</artifactId>
            <version>3.16.1</version>
            <scope>provided</scope>
        </dependency>
```

ckunki commented on pom.xml:
```diff
@@ -0,0 +1,190 @@
+            <version>3.16.1</version>
+        </dependency>
+        <!--Unit test dependencies -->
+        <dependency>
```
Proposal: move  virtual-schema-common-jdbc test-jar below comment  `<!-- Integration test dependencies -->`

pj-spoelders replied:
Done

ckunki commented on pom.xml:
```diff
@@ -0,0 +1,190 @@
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.hamcrest</groupId>
```
A newer version of org.hamcrest is available. 
Proposal:  run `mvn versions:use-latest-releases`

pj-spoelders replied:
Maybe in a future maintenance release. I'm not touching the pom anymore for now.

ckunki commented on pom.xml:
```diff
@@ -0,0 +1,190 @@
+            <artifactId>localstack</artifactId>
+            <version>1.19.8</version>
+            <scope>test</scope>
+        </dependency>
```
```suggestion
```

pj-spoelders replied:
Makes sense.

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeColumnMetadataReader.java:
```diff
@@ -0,0 +1,77 @@
+        super(connection, properties, identifierConverter);
+    }
+
+    /**
+     * Get the catalog name that is applied as filter criteria when looking up remote metadata.
+     *
+     * @return catalog name or <code>null</code> if metadata lookups are not limited by catalog
+     */
```
```suggestion
```

pj-spoelders replied:
I don't see what you mean?

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeColumnMetadataReader.java:
```diff
@@ -0,0 +1,77 @@
+        }
+    }
+
+    protected DataType mapJdbcTypeOther(final JDBCTypeDescription jdbcTypeDescription) {
```
```suggestion
    private DataType mapJdbcTypeOther(final JDBCTypeDescription jdbcTypeDescription) {
```

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeColumnMetadataReader.java:
```diff
@@ -0,0 +1,77 @@
+    protected DataType mapJdbcTypeOther(final JDBCTypeDescription jdbcTypeDescription) {
+        if (isVarBitColumn(jdbcTypeDescription)) {
+            final int n = jdbcTypeDescription.getPrecisionOrSize();
+            LOGGER.finer(() -> "Mapping Snowflake datatype \"OTHER:varbit\" to VARCHAR(" + n + ")");
```
```suggestion
            LOGGER.finest(() -> "Mapping Snowflake datatype 'OTHER:varbit' to VARCHAR(" + n + ")");
```

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeColumnMetadataReader.java:
```diff
@@ -0,0 +1,77 @@
+            return DataType.createVarChar(n, DataType.ExaCharset.UTF8);
+        } else {
+            LOGGER.finer(() -> "Mapping Snowflake datatype \"" + jdbcTypeDescription.getTypeName()
+                    + "\" to maximum VARCHAR()");
```
```suggestion
            LOGGER.finest(() -> "Mapping Snowflake datatype '" + jdbcTypeDescription.getTypeName()
                    + "' to maximum VARCHAR()");
```

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeColumnMetadataReader.java:
```diff
@@ -0,0 +1,77 @@
+        }
+    }
+
+    protected boolean isVarBitColumn(final JDBCTypeDescription jdbcTypeDescription) {
```
```suggestion
    private boolean isVarBitColumn(final JDBCTypeDescription jdbcTypeDescription) {
```

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeColumnMetadataReader.java:
```diff
@@ -0,0 +1,77 @@
+    @Override
+    public String readColumnName(final ResultSet columns) throws SQLException {
+            return super.readColumnName(columns);
+    }
```
```suggestion
    public String readColumnName(final ResultSet columns) throws SQLException {
        return super.readColumnName(columns);
    }
```

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeMetadataReader.java:
```diff
@@ -0,0 +1,45 @@
+    public String getSchemaNameFilter() {
+        return this.properties.getSchemaName().replace("_","\\_").toUpperCase();
+    }
+
```
```suggestion
```

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialect.java:
```diff
@@ -0,0 +1,196 @@
+
+    private static final Set<ScalarFunctionCapability> DISABLED_SCALAR_FUNCTION = Set.of(
+            /*
+             * Implementation for `BETWEEN` time functions is not supported. For more information see `design.md` file,
```
Proposal: `BETWEEN` time functions are not supported.

File `design.md` is empty. Please add details.

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialect.java:
```diff
@@ -0,0 +1,196 @@
+
+    /*
+     * IMPORTANT! Before adding new capabilities, check the `doc/design.md` file if there is a note on why the
+     * capability is not supported. it.
```
File `design.md` is empty. Please add details.
```suggestion
     * IMPORTANT! Before adding new capabilities, check the `doc/design.md` file if there is a note on why the
     * capability is not supported. 
```

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialect.java:
```diff
@@ -0,0 +1,196 @@
+
+    /**
+     * This class gets all {@link ScalarFunctionCapability}s that are not explicitly excluded by
+     * {@link #DISABLED_SCALAR_FUNCTION}.
```
```suggestion
     * Return all {@link ScalarFunctionCapability}s that are not explicitly excluded by
     * {@link #DISABLED_SCALAR_FUNCTION}.
```

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialect.java:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    @Override
+
```
```suggestion
```

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlGenerationVisitor.java:
```diff
@@ -0,0 +1,225 @@
+        super(dialect, context);
+    }
+
+    protected List<String> getListOfTypeNamesRequiringCast() {
```
Replace `protected` by `private` or maybe annotation `@Override` is missing?

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlGenerationVisitor.java:
```diff
@@ -0,0 +1,225 @@
+        return TYPE_NAMES_REQUIRING_CAST;
+    }
+
+    protected List<String> getListOfTypeNamesNotSupported() {
```
Replace `protected` by `private` or maybe annotation `@Override` is missing?

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlGenerationVisitor.java:
```diff
@@ -0,0 +1,225 @@
+
+    private String getColumnProjectionString(final SqlColumn column, final String projectionString)
+            throws AdapterException {
+
```
```suggestion
```

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlGenerationVisitor.java:
```diff
@@ -0,0 +1,225 @@
+            ColumnAdapterNotes columnAdapterNotes= converter.convertFromJsonToColumnAdapterNotes(metaData.getAdapterNotes(), column.getName());
+                    String sourceTypeName = columnAdapterNotes.getTypeName();
+
+                return buildColumnProjectionString(sourceTypeName,mappedType, projectionString); //
```
```suggestion
            return buildColumnProjectionString(sourceTypeName,mappedType, projectionString); 
```

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlGenerationVisitor.java:
```diff
@@ -0,0 +1,225 @@
+            ColumnMetadata metaData = column.getMetadata();
+            DataType mappedType = metaData.getType();
+            ColumnAdapterNotes columnAdapterNotes= converter.convertFromJsonToColumnAdapterNotes(metaData.getAdapterNotes(), column.getName());
+                    String sourceTypeName = columnAdapterNotes.getTypeName();
```
```suggestion
            String sourceTypeName = columnAdapterNotes.getTypeName();
```

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlGenerationVisitor.java:
```diff
@@ -0,0 +1,225 @@
+
+                return buildColumnProjectionString(sourceTypeName,mappedType, projectionString); //
+        } else {
+                return projectionString;
```
```suggestion
            return projectionString;
```

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlGenerationVisitor.java:
```diff
@@ -0,0 +1,225 @@
+        } else if (typeName.startsWith("bigserial")) {
+            return "CAST(" + projectionString + "  as BIGINT )";
+        } else if (TYPE_NAMES_NOT_SUPPORTED.contains(typeName)) {
+            return "cast('" + typeName + " NOT SUPPORTED' as varchar) as not_supported";
```
Would it be better to throw an exception here?

pj-spoelders replied:
Removed this code. There's no serial types in snowflake.

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlGenerationVisitor.java:
```diff
@@ -0,0 +1,225 @@
+
+    private String buildColumnProjectionString(final String typeName,DataType mappedType, final String projectionString) {
+        if (typeName.startsWith("NUMBER") && mappedType.getExaDataType() == DataType.ExaDataType.VARCHAR ){
+            return "'Number precision not supported'";
```
Would it be better to throw an exception here?

pj-spoelders replied:
No, this is what was decided in the meeting.

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlGenerationVisitor.java:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    private boolean checkIfNeedToCastToVarchar(final String typeName) {
+        final List<String> typesToVarcharCast = Arrays.asList("point", "line", "varbit", "lseg", "box", "path",
```
Proposal: create a constant and use a set
```suggestion
        private static final Set<String> TYPES_TO_VARCHAR_CAST = Set.of("point", "line", "varbit", "lseg", "box", "path",
```

pj-spoelders replied:
I won't touch this, I feel it doesn't matter, it's test code ..

ckunki commented on src/main/java/com/exasol/adapter/dialects/snowflake/SnowflakeTableMetadataReader.java:
```diff
@@ -0,0 +1,31 @@
+ * This class handles the specifics of mapping Snowflake table metadata to Exasol.
+ */
+public class SnowflakeTableMetadataReader extends BaseTableMetadataReader {
+    static final Logger LOGGER = Logger.getLogger(SnowflakeTableMetadataReader.class.getName());
```
```suggestion
```

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeColumnMetadataReaderTest.java:
```diff
@@ -0,0 +1,58 @@
+
+class SnowflakeColumnMetadataReaderTest {
+    private SnowflakeColumnMetadataReader columnMetadataReader;
+    private Map<String, String> rawProperties;
```
```suggestion
```

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeColumnMetadataReaderTest.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+
+
```
```suggestion
```

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialectIT.java:
```diff
@@ -0,0 +1,373 @@
+        try {
+            statementSnowflake.execute("CREATE DATABASE " + SETUP.getDatabaseName());
+        } catch (final Exception e) {
+
```
Please add code handling the exception.
Probably the exception is thrown (and ignored) when the database already exists. 
In this case the setup needs to delete the database before continuing.

pj-spoelders replied:
Removed the try catch since it isn't needed anymore with the newer test setup code.

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialectIT.java:
```diff
@@ -0,0 +1,373 @@
+            statementSnowflake.execute("DROP SCHEMA " + SCHEMA_SNOWFLAKE + " CASCADE");
+            statementSnowflake.execute("DROP SCHEMA " + SCHEMA_SNOWFLAKE_UPPERCASE_TABLE + " CASCADE");
+        } catch (final Exception e) {
+
```
Please add code handling the exception.

pj-spoelders replied:
Removed the try/catch clause instead. Made changes to do this in a different way (Whole database gets dropped now in AfterAll)

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialectIT.java:
```diff
@@ -0,0 +1,373 @@
+        try {
+            statementSnowflake.execute("DROP DATABASE " + SETUP.getDatabaseName() + " CASCADE;");
+        } catch (final Exception e) {
+
```
Please add code handling the exception.

pj-spoelders replied:
No need for try/catch clause. Removed the try catch statement.

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialectIT.java:
```diff
@@ -0,0 +1,373 @@
+
+    @Test
+    void testSelectSingleColumn() throws SQLException {
+        final ResultSet actualResultSet = statementExasol
```
Proposal: create a private method accepting the matcher as parameter and doing
1. execute statement 
2. get result set 
3. verify result 
4. close result set

pj-spoelders replied:
I'm not reworking the tests.

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialectIT.java:
```diff
@@ -0,0 +1,373 @@
+                List.of("x  DECIMAL(36,0)", "y VARCHAR(100)", "a DECIMAL(36,0)", "b VARCHAR(100)"), //
+                List.of("2,'bbb', 2,'bbb'"));
+        final ResultSet actualResultSet = getActualResultSet(query);
+        assertThat(actualResultSet, matchesResultSet(expected));
```
Proposal: Use `ResultSetMatcher` as in `testSelectSingleColumn()`  

Alternatively in method `getExpectedResultSet()` do not create a physical SQL table but use a select statement `FROM VALUES`.

pj-spoelders replied:
No. I'm not altering these tests any further.

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialectIT.java:
```diff
@@ -0,0 +1,373 @@
+        final String query = "SELECT year(\"MYDATE\") FROM " + virtualSchemaSnowflake.getName() + "."
+                + TABLE_SNOWFLAKE_ALL_DATA_TYPES;
+        final ResultSet actualResultSet = getActualResultSet(query);
+        final Short yearShort = 2010;
+        assertThat(actualResultSet, table().row(yearShort).matches());
```
```suggestion
        assertThat(actualResultSet, table().row((short) 2010).matches());
```

pj-spoelders replied:
Not touching this.

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialectIT.java:
```diff
@@ -0,0 +1,373 @@
+        MatcherAssert.assertThat(actualResultSet, table().row(expectedValue).matches(TypeMatchMode.NO_JAVA_TYPE_CHECK));
+    }
+
+    // TODO refactor to use table().row().matches()
```
Please implement this TODO.

pj-spoelders replied:
Out of scope, it'll have to wait.

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialectIT.java:
```diff
@@ -0,0 +1,373 @@
+    private ResultSet getActualResultSet(final String query) throws SQLException {
+        return statementExasol.executeQuery(query);
+    }
+
```
```suggestion
```

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialectTest.java:
```diff
@@ -0,0 +1,127 @@
+        );
+    }
+
+    @ValueSource(strings = { "ab:E'ab'", "a'b:E'a''b'", "a''b:E'a''''b'", "'ab':E'''ab'''", "a\\\\b:E'a\\\\\\\\b'",
```
Proposal: use `String.split()` to split the definition.

pj-spoelders replied:
Not touching these.

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialectTest.java:
```diff
@@ -0,0 +1,127 @@
+    void testGetLiteralStringNull() {
+        assertThat(this.dialect.getStringLiteral(null), CoreMatchers.equalTo("NULL"));
+    }
+
```
```suggestion
```

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialectTest.java:
```diff
@@ -0,0 +1,127 @@
+        final SqlDialect sqlDialect = new SnowflakeSqlDialect(null, new AdapterProperties(Map.of( //
+                CONNECTION_NAME_PROPERTY, "MY_CONN", //
+                CATALOG_NAME_PROPERTY, "TESTDB")));
+        sqlDialect.validateProperties();
```
Proposal: use `assertDoesNotThrow()`

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlDialectTest.java:
```diff
@@ -0,0 +1,127 @@
+                SCHEMA_NAME_PROPERTY, "MY_SCHEMA")));
+        sqlDialect.validateProperties();
+    }
+}
```
Please add test verifying an exception is thrown in case of illegal properties, illegal property values or missing mandatory properties.

pj-spoelders replied:
I've made a ticket about adding SIT tests to the VS. I suggest we make these suggestions and/or changes part of this PR.

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlGenerationVisitorTest.java:
```diff
@@ -0,0 +1,107 @@
+            "WEEK, WEEK, 2", //
+            "MONTH, MONTH, 2", //
+            "YEAR, YEAR, 4" })
+    @ParameterizedTest
```
Please add test for unsupported scalar function, e.g. `SIN`.

pj-spoelders replied:
I've made a ticket about adding SIT tests to the VS. I suggest we make these suggestions and/or changes part of this PR.

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeSqlGenerationVisitorTest.java:
```diff
@@ -0,0 +1,107 @@
+            "ADD_WEEKS, weeks", //
+            "ADD_MONTHS, months" })
+    @ParameterizedTest
+    void testVisitSqlFunctionScalarAddDate(final ScalarFunction scalarFunction, final String expected)
```
Please add test for unsupported scalar function, e.g. `SIN`.

pj-spoelders replied:
I've made a ticket about adding SIT tests to the VS. I suggest we make these suggestions and/or changes part of this PR.

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeTableMetadataReaderTest.java:
```diff
@@ -0,0 +1,35 @@
+class SnowflakeTableMetadataReaderTest {
+    private Map<String, String> rawProperties;
+    private SnowflakeTableMetadataReader reader;
+
```
Please add tests.

pj-spoelders replied:
I've made a ticket about adding SIT tests to the VS. I suggest we make these suggestions and/or changes part of this PR.

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeVirtualSchemaIntegrationTestSetup.java:
```diff
@@ -0,0 +1,252 @@
+    private static final Path PATH_TO_VIRTUAL_SCHEMAS_JAR = Path.of("target", VIRTUAL_SCHEMAS_JAR_NAME_AND_VERSION);
+    private static final String SCHEMA_EXASOL = "SCHEMA_EXASOL";
+    private static final String ADAPTER_SCRIPT_EXASOL = "ADAPTER_SCRIPT_EXASOL";
+    private static final String EXASOL_DOCKER_IMAGE_REFERENCE = "8.24.0";
```
```suggestion
    private static final String EXASOL_DOCKER_IMAGE_REFERENCE = "8.31.0";
```

pj-spoelders replied:
Done

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeVirtualSchemaIntegrationTestSetup.java:
```diff
@@ -0,0 +1,252 @@
+    private static final String SCHEMA_EXASOL = "SCHEMA_EXASOL";
+    private static final String ADAPTER_SCRIPT_EXASOL = "ADAPTER_SCRIPT_EXASOL";
+    private static final String EXASOL_DOCKER_IMAGE_REFERENCE = "8.24.0";
+    private static final String SNOWFLAKE_CONTAINER_NAME = "localstack/snowflake:14.2";
```
unused 
```suggestion
```

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeVirtualSchemaIntegrationTestSetup.java:
```diff
@@ -0,0 +1,252 @@
+    private static final String JDBC_DRIVER_NAME = "snowflake-jdbc.jar";
+    private static final Path JDBC_DRIVER_PATH = Path.of("target/snowflake-driver/" + JDBC_DRIVER_NAME);
+
+    private static final int SNOWFLAKE_PORT = 5432;
```
```suggestion
```

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeVirtualSchemaIntegrationTestSetup.java:
```diff
@@ -0,0 +1,252 @@
+    private static final int SNOWFLAKE_PORT = 5432;
+    private static final String USERNAME_FILE = "username.txt";
+    private static final String PASSWORD_FILE = "password.txt";
+    private static final String ACCOUNTNAME_FILE = "accountname.txt";
```
Proposal: use properties file, see https://github.com/exasol/databricks-virtual-schema/blob/main/src/test/java/com/exasol/adapter/databricks/fixture/TestConfig.java

pj-spoelders replied:
👍 done!

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeVirtualSchemaIntegrationTestSetup.java:
```diff
@@ -0,0 +1,252 @@
+        final String generatedString = random.ints(leftLimit, rightLimit + 1).limit(targetStringLength)
+                .collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append).toString();
+
+        return generatedString;
```
```suggestion
        return random.ints(leftLimit, rightLimit + 1).limit(targetStringLength)
                .collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append).toString();
```

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeVirtualSchemaIntegrationTestSetup.java:
```diff
@@ -0,0 +1,252 @@
+
+    private Connection getSnowflakeConnection(final String username, final String password, final String accountname)
+            throws SQLException {
+        try {
+            Class.forName("net.snowflake.client.jdbc.SnowflakeDriver");
+        } catch (final ClassNotFoundException ex) {
+            System.err.println("Driver not found");
+        }
```
Obsolete - class will be loaded automatically.
But if loading explicitly, then just let the exception be raised, don't catch it.

pj-spoelders replied:
Removed try catch and let it bubble up to general snowflake setup failure handling if it fails.

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeVirtualSchemaIntegrationTestSetup.java:
```diff
@@ -0,0 +1,252 @@
+        properties.put("password", password); // replace "" with your password
+        properties.put("account", accountname); // replace "" with your account name
+        properties.put("db", this.databaseName); // replace "" with target database name
+        properties.put("schema", "TESTSCHEMA"); // replace "" with target schema name
```
```suggestion
        properties.put("user", username); 
        properties.put("password", password); 
        properties.put("account", accountname); 
        properties.put("db", this.databaseName); 
        properties.put("schema", "TESTSCHEMA"); 
```

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeVirtualSchemaIntegrationTestSetup.java:
```diff
@@ -0,0 +1,252 @@
+        properties.put("account", accountname); // replace "" with your account name
+        properties.put("db", this.databaseName); // replace "" with target database name
+        properties.put("schema", "TESTSCHEMA"); // replace "" with target schema name
+        // properties.put("tracing", "on");
```
```suggestion
```

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeVirtualSchemaIntegrationTestSetup.java:
```diff
@@ -0,0 +1,252 @@
+        // use the default connection string if it is not set in environment
+        if (connectStr == null) {
+            connectStr = "jdbc:snowflake://" + accountname + ".snowflakecomputing.com"; // replace accountName with your account name
+        }
```
```suggestion
       String connectStr = "jdbc:snowflake://" + accountname + ".snowflakecomputing.com"; 
```

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeVirtualSchemaIntegrationTestSetup.java:
```diff
@@ -0,0 +1,252 @@
+        return DriverManager.getConnection(connectStr, properties);
+    }
+
+    private void getTestCredentials() throws IOException {
```
See above, use properties.

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeVirtualSchemaIntegrationTestSetup.java:
```diff
@@ -0,0 +1,252 @@
+    }
+
+    private AdapterScript createAdapterScript(final ExasolSchema schema) {
+        final String content = "%scriptclass com.exasol.adapter.RequestDispatcher;\n" //
```
TODO: Add path to JDBC driver, as this is not included in the adapter jar anymore (changed scope to "provided" in pom file).

pj-spoelders replied:
I'm not sure why, but this seems to work fine for the integration tests. It's the same way here:
https://github.com/exasol/postgresql-virtual-schema/blob/ff7bfe9e0358862adc030d6cb9a6b2fc9120290c/src/test/java/com/exasol/adapter/dialects/postgresql/PostgresVirtualSchemaIntegrationTestSetup.java#L120
Any ideas as to why?

ckunki replied:
https://github.com/exasol/oracle-virtual-schema/blob/main/src/test/java/com/exasol/adapter/dialects/oracle/OracleVirtualSchemaIntegrationTestSetup.java#L87

pj-spoelders replied:
As discussed, will not touch for now. It remains a 'mystery', (there's no documentation available that talks about this in details) for now and we should probably investigate when we have more time or ask around, preferably to one of the original implementers.

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeVirtualSchemaIntegrationTestSetup.java:
```diff
@@ -0,0 +1,252 @@
+
+    public VirtualSchema createVirtualSchema(final String forSnowflakeSchema,
+                                             final Map<String, String> additionalProperties) {
+        // TODO add localstack support + cleanup
```
```suggestion
```

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeVirtualSchemaIntegrationTestSetup.java:
```diff
@@ -0,0 +1,252 @@
+            this.snowflakeStatement.close();
+            this.snowflakeConnection.close();
+            this.exasolContainer.stop();
+            // TODO add localstack support + cleanup
+            // this.snowflakeContainer.stop();
```
```suggestion
```

ckunki commented on src/test/java/com/exasol/adapter/dialects/snowflake/SnowflakeVirtualSchemaIntegrationTestSetup.java:
```diff
@@ -0,0 +1,252 @@
+        final Map<String, ContainerNetwork> networks = this.exasolContainer.getContainerInfo().getNetworkSettings()
+                .getNetworks();
+        if (networks.size() == 0) {
+            return null;
```
Throw an exception here instead of returning null.

pj-spoelders replied:
Any suggestions? Also, there's a assertion after this function call to check it's not null.

pj-spoelders replied:
Ok, since there's an assert, we shouldn't bother, as discussed.

ckunki reviewed (commented):
See comments

kaklakariada reviewed (changes requested):
Please see comments

pj-spoelders committed:
PR review feedback (1)
-  update .gitattributes
- remove stale todo comments
- update exasolDbversions in project-keeper.yml

pj-spoelders committed:
PR review feedback (2)
- Use TestConfig + test.properties file instead of 3 different files for accountname, username & password

pj-spoelders committed:
change release date

pj-spoelders committed:
Update copyright license

pj-spoelders committed:
pk dependency update fix

pj-spoelders committed:
fix changes file

pj-spoelders committed:
update testcontainers to also work with testcontainer versions > 8.29.0 (fix nullpointer exception)

kaklakariada committed:
Fix compile error

pj-spoelders committed:
cleanup code smells after testconfig refactor

pj-spoelders committed:
release date

pj-spoelders committed:
fix compilation error after refactoring

pj-spoelders committed:
reformatted  SnowflakeSqlGenerationVisitor

pj-spoelders committed:
set docker image reference in integration test setup to latest dockerdb version

pj-spoelders committed:
removed unused dependency + moved dependency to integration test dependency block

pj-spoelders committed:
cleaned up IntegrationTestSetup class

pj-spoelders committed:
removed //

pj-spoelders committed:
simplified snowflake database setup and teardown in IT class

pj-spoelders committed:
PR Review: requested changes to user guide

pj-spoelders committed:
removed 'serial' cases in visitor since these don't exist in snowflake

pj-spoelders committed:
fix formatting

pj-spoelders committed:
changes to visibility and logger settings

pj-spoelders committed:
changes to exception handling

pj-spoelders committed:
Removed unnecessary whitespace

pj-spoelders committed:
Changed comment as suggested

pj-spoelders committed:
user guide, redid conversion/mapping table

pj-spoelders committed:
design and dev guide

pj-spoelders committed:
Changes to readme.md

pj-spoelders committed:
Note on deviations

pj-spoelders committed:
Update settings.json to remove debug status and use latest docker image

pj-spoelders committed:
Removed note on scalar functions

pj-spoelders committed:
Removed unused localstack dependency, PK update to changes file

pj-spoelders committed:
Removed newline

pj-spoelders committed:
PK/dependencies/doc fix

pj-spoelders committed:
removed unused code

pj-spoelders committed:
release date

kaklakariada commented on doc/changes/changes_0.1.0.md:
```diff
@@ -1,4 +1,4 @@
-# Virtual Schema for Snowflake 0.1.0, released 2024-09-26
+# Virtual Schema for Snowflake 0.1.0, released 2024-09-27
```
```suggestion
# Virtual Schema for Snowflake 0.1.0, released 2024-10-01
```

pj-spoelders committed:
fix dependencies in changes doc
